File: src/main/programs/array/ProductOfArrayExceptSelf.java
package array;

import java.util.Arrays;

public class ProductOfArrayExceptSelf {
    public static int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];

        // Initialize result array with 1s
        result[0] = 1;

        // Calculate left products
        for (int i = 1; i < n; i++) {
            result[i] = result[i-1] * nums[i-1];
        }

        // Calculate right products and combine
        int rightProduct = 1;
        for (int i = n-1; i >= 0; i--) {
            result[i] *= rightProduct;
            rightProduct *= nums[i];
        }

        return result;
    }

    public static void main(String[] args) {

        int[] nums = {1, 2, 3, 4};
        int[] result = productExceptSelf(nums);
        System.out.println("Result: " + Arrays.toString(result));
    }
}

File: src/main/programs/array/ProductOfArrayExceptSelf.kt
package array

//Using left and right products (most efficient)
fun productOfArrayExceptSelf(nums:IntArray):IntArray{
    val n=nums.size
    val result=IntArray(n)
    result[0]=1

    for (i in 1 until n){
        result[i]=result[i-1] * nums[i-1]
    }

    var righProduct=1
    for (i in n-1 downTo 0){
        result[i] *=righProduct
        righProduct *=nums[i]
    }
    return result
}

fun productOfArrayExceptSelfFunctional(nums: IntArray):IntArray{
    return nums.indices.map { i->
        nums.filterIndexed { index, _ -> index!=i  }
            .fold(1){acc, num -> acc*num  }
    }.toIntArray()
}

fun productExceptSelf(nums: IntArray): IntArray {
    val n = nums.size
    val leftProducts = IntArray(n) { 1 }
    val rightProducts = IntArray(n) { 1 }
    val result = IntArray(n)

    // Calculate left products
    for (i in 1 until n) {
        leftProducts[i] = leftProducts[i-1] * nums[i-1]
    }

    // Calculate right products
    for (i in n-2 downTo 0) {
        rightProducts[i] = rightProducts[i+1] * nums[i+1]
    }

    // Combine left and right products
    for (i in 0 until n) {
        result[i] = leftProducts[i] * rightProducts[i]
    }

    return result
}

fun main(){
    val nums = intArrayOf(1, 2, 3, 4)
    val result = productExceptSelf(nums)
    println("Result: ${result.contentToString()}")
}